File: bubble-sort/code.cpp
#include <iostream>

//Simple bubble sort
void bubble_sort(int array[] , int size){
   for(int i = 0; i < size - 1; i++){

      for(int j = 0; j < size - 1 - i; j++){

         if(array[j] > array[j+1]){
            //swap elements
            int x = array[j];
            array[j] = array[j + 1];
            array[j + 1] = x;

         }

      }
   }
}

//Adaptive Bubble Sort
void adaptive_bubble_sort(int array[] , int size){
   int flag;
   
   for(int i = 0; i < size -1 ; i++){

      flag = 0;
      for(int j = 0; j < size - 1 - i; j++){

         
         if(array[j] > array[j + 1]){
            //swap
            int x = array[j];
            array[j] = array[j + 1];
            array[j + 1] = x;

            //increment the flag
            flag++;

         }

      }

      //Check if swapping wasnt done, If no swapping means array is already sorted so break
      
      if(flag == 0){
         break;
      }
   }

}

//Function that prints the array
void print_array(int array[] , int size){
   for(int i = 0; i < size; i++){
      std::cout<< array[i] << ",";
   }
}

int main(){

   int unordered_array[] = {8,2,3,4,5};
   int size = 5;
   
   adaptive_bubble_sort(unordered_array, size);

   //printing the array
   print_array(unordered_array, size);
   return 0;
}

File: bubble-sort/code.py
# Book Implementation:
def bubble_sort(list ):
   sorted_until_index = len(list) - 1
   sorted = False

   while not sorted:
      sorted = True

      for i in range(sorted_until_index):

         if list[i] > list[i + 1]:
            sorted = False
            #swap
            list[i] , list[i + 1] = list[i + 1] , list[i]

      #Last value is sorted
      sorted_until_index -= 1

#Implementation 2
def bubble_sort_2(list):
   for i in range(0 , len(list) - 1):
      flag = 0

      for j in range(0 , len(list) - 1 - i):

         if list[j] > list[j + 1]:
            #swap
            list[j] , list[j + 1] = list[j + 1] , list[j]
            flag += 1
         
      if flag == 0:
         break

list = [8,5,3,2,7]
bubble_sort_2(list)
print(list)